### ISSUE TITLE
Question: Opensource modulesystem

### ISSUE THREAD

**Issue by wardpeet** [callstack/linaria#201]
I really like the [module system](https://github.com/callstack/linaria/blob/master/src/babel/lib/moduleSystem.js) as it is great to compile modules and generate the correct output.

I really want to use it in [emotion](https://github.com/emotion-js/emotion/tree/master/packages/babel-plugin-emotion)

So my request is if you could create a separate package for this one. For now I can start copying the code but I would love to just include it as a package as it's easier to give credits where credits are due.

**Comment by satya164** [comment]
Hey @wardpeet! Really interested in knowing what you're going to use it for :D

**Comment by jacobp100** [comment]
Is this the part that does build-time code evaluation? I want to use it in [cssta](https://github.com/jacobp100/cssta) (the web version). It looks like we have really similar projects---I wonder what overlap we could share!

**Comment by wardpeet** [comment]
@satya164 also css interpolation. Emotion babel support has the same option to extractCSS to a cssFile. The downsize is that it doesn't support interpolation which is sad. I'll use the module compiler to resolve my ast to code like you do with linaria

```
var MY_CONST = '#000';

const myClass = css`
  background: ${MY_CONST};
`;
```

```
const myClass = css`
  background: #000;
`;
```

**Comment by zamotany** [comment]
@wardpeet 2 things:

1. Why don't you use linaria?
2. The module system is not responsible for evaluation at compile time.

**Comment by wardpeet** [comment]
@zamotany syntax is a bit of with the include statement. We really like the styled components syntax but we need the css extraction for speed and emotion was the only option.

We would use it the same as the lernia babel plugin so not doing anything on compile time

**Comment by jacobp100** [comment]
I think what we would want is `babel-plugin-eval-template-expressions`, which takes options like `{ templateTag: 'glamor' }`. This would then be able to be used in multiple CSS-in-JS projects, which would be great for collaboration 😄

**Comment by satya164** [comment]
> syntax is a bit of with the include statement

`include` is just a helper to expand a class name to get the styles. emotion has a similar feature where you can write `${someClass}` instead of `${include(someClass)}` which is confusing imo. What if you actually want to use the class name in a selector like `& > .${someClass}`? Now the same syntax behaves completely different according to where it's used. That said, `include` is not something that you use frequently.

Regarding a syntax like styled components, we are thinking about that: https://gist.github.com/satya164/a622deac263f48632ad149d73f1bb229

> I think what we would want is babel-plugin-eval-template-expressions, which takes options like { templateTag: 'glamor' }.

That'll be nice. But I don't think we currently have resources to spend time on doing this. If you wanna do it, go ahead!

**Comment by jacobp100** [comment]
I made a really _really_ basic start on this over [here](https://github.com/jacobp100/babel-plugin-pre-evaluate-template-strings). For now, I've just copied and pasted code.

I suspect we'll first need to expose the `resolveTemplateLiteral` function. We'll also need ways to customise the interpolation too. We might need to expose module mocking.

The aim is that any CSS-in-JS project should be able to implement it, and if they don't, a user should be able to use the babel plugin directly. It's a start at least!
